### PR TITLE
Remove Jersey dependencies from Eureka Client

### PR DESCRIPTION
**Pull request by daniellavoie**
Fixes gh-1849

This is a first draft. I am going to need some help since I did not change the test coverage. The new RestTemplate implementation passes all tests but I am unaware of the coverage of all the existing tests. Theoritically, new tests should not be required if the existing tests are complete.

Note that this change does not remove the dependencies of Jersey from Ribbon. 

There are so much subtleties with Eureka that I am going to need a solid code review to make sure that I did not forget something.

This PR does not provide any new feature so I assumed that existing tests should be enough.

### CODE REVIEW

**Comment by daniellavoie** [comment]
Alright, I guess I can do better with coverage. I'll come back with improvements.

File: spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/EurekaClientAutoConfiguration.java [hunk 1]
```diff
@@ -212,18 +203,14 @@ public ApplicationInfoManager eurekaApplicationInfoManager(
 		@Autowired
 		private ApplicationContext context;
 
-		@Autowired(required = false)
-		private DiscoveryClientOptionalArgs optionalArgs;
```
Comment (spencergibb) [inline spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/EurekaClientAutoConfiguration.java]: We have users who rely on this BTW.

File: spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/http/RestTemplateDiscoveryClientOptionalArgs.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+import com.netflix.discovery.AbstractDiscoveryClientOptionalArgs;
+
+/**
+ * @author Daniel Lavoie
+ */
+public class RestTemplateDiscoveryClientOptionalArgs
+		extends AbstractDiscoveryClientOptionalArgs<Void> {
+	public RestTemplateDiscoveryClientOptionalArgs() {
+		setTransportClientFactories(new RestTemplateTransportClientFactories());
```
Comment (spencergibb) [inline spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/http/RestTemplateDiscoveryClientOptionalArgs.java]: I'm wondering if this should be opt in for the first release train, then the default later. @dsyer?

File: spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/http/RestTemplateEurekaHttpClient.java [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+	private String serviceUrl;
+
+	public RestTemplateEurekaHttpClient(RestTemplate restTemplate, String serviceUrl) {
+		this.restTemplate = restTemplate;
+		this.serviceUrl = serviceUrl;
+	}
+
+	@Override
+	public EurekaHttpResponse<Void> register(InstanceInfo info) {
+		String urlPath = serviceUrl + "apps/" + info.getAppName();
```
Comment (spencergibb) [inline spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/http/RestTemplateEurekaHttpClient.java]: I assume you translated from the Jersey classes?

File: spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/EurekaClientAutoConfiguration.java (hunk 1 above)
Comment (daniellavoie) [inline spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/EurekaClientAutoConfiguration.java]: I was affraid of that one. The only workaound I can think of is providing an abstraction of `DiscoveryClientOptionalArgs` which will be compatible with injection without forcing the dependency on Jersey. I'll rework that bit.

File: spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/http/RestTemplateEurekaHttpClient.java (hunk 3 above)
Comment (daniellavoie) [inline spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/http/RestTemplateEurekaHttpClient.java]: This is basically a dumb rewrite of [AbstractJersey2EurekaHttpClient](https://github.com/Netflix/eureka/blob/master/eureka-client/src/main/java/com/netflix/discovery/shared/transport/jersey/AbstractJerseyEurekaHttpClient.java) with a RestTemplate. I'm not sure Spring Cloud Eureka client uses all the features of the client. I'm going to need some insight.

File: spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/http/RestTemplateDiscoveryClientOptionalArgs.java (hunk 2 above)
Comment (daniellavoie) [inline spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/http/RestTemplateDiscoveryClientOptionalArgs.java]: I'm confortable with a opt-in considering ribbon still pulls Jersey. A first step with opt in could road test our standalone http client for people who don't need ribbon.

**Comment by daniellavoie** [comment]
I have contributed new changes for a opt-in approach. The new `RestTemplateTransportClientFactories` will active itself only if the jersey dependencies are explicitely exclude from the `spring-cloud-starter-eureka`. The `DiscoveryClientOptionalArgs` being provided by clients is still supported. 

Last thing I need to work on is a better code coverage for `RestTemplateEurekaHttpClient`.

File: spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/EurekaClientAutoConfiguration.java [hunk 4]
```diff
@@ -114,9 +118,11 @@ public EurekaClientConfigBean eurekaClientConfigBean() {
 	@Bean
 	@ConditionalOnMissingBean(value = EurekaInstanceConfig.class, search = SearchStrategy.CURRENT)
 	public EurekaInstanceConfigBean eurekaInstanceConfigBean(InetUtils inetUtils) {
-		RelaxedPropertyResolver relaxedPropertyResolver = new RelaxedPropertyResolver(env, "eureka.instance.");
+		RelaxedPropertyResolver relaxedPropertyResolver = new RelaxedPropertyResolver(env,
```
Comment (spencergibb) [inline spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/EurekaClientAutoConfiguration.java]: Can you refrain from unrelated formatting? Makes it hard to see what you are actually changing.

Comment (daniellavoie) [inline spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/EurekaClientAutoConfiguration.java]: That one is is undesired, my bad. BTW thanks for the filtered classpath trick, I'll look into it.

**Comment by daniellavoie** [comment]
@ryanjbaxter do you have any idea why my latest build did not updated the codecov report ?

**Comment by daniellavoie** [comment]
Just contributed an update that should address all pending issues. I ended up building a mocked Eureka Server. Code coverage as increased.

**Comment by daniellavoie** [comment]
@spencergibb Any other remarks ?

**Comment by spencergibb** [comment]
Targeting the Edgware release this summer, should be merged in the next week or two.
